### PR TITLE
update trace_filter, trace block only to highest matching index

### PR DESCRIPTION
**Pull request by smatthewenglish**
This PR aims to address the following TODO:
```
// TODO: this could be optimized to only trace the block until the highest matching index in
// that block
```

This is my first contribution to `reth`, I've read through the CONTRIBUTING.md doc and executed all the specified checks. If there's anything I've overlooked or missed please let me know and I'll be happy to address it.

### CODE REVIEW

File: crates/rpc/rpc/src/trace.rs [hunk 1]
```diff
@@ -276,29 +276,34 @@ where
...
             for (tx_idx, tx) in block.body.iter().enumerate() {
                 let from = tx.recover_signer().ok_or(BlockError::InvalidSignature)?;
                 let to = tx.to();
                 if matcher.matches(from, to) {
-                    transaction_indices.insert(tx_idx as u64);
+                    let idx = tx_idx as u64;
+                    transaction_indices.insert(idx);
+                    if idx > highest_matching_index {
+                        highest_matching_index = idx;
+                    }
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:288]: this check is redundant because all tx_idx are only increasing

File: crates/rpc/rpc/src/trace.rs [hunk 2]
```diff
@@ -276,29 +276,34 @@ where
...
-        for (num, indices) in target_blocks {
+        for (num, indices, highest_idx) in target_blocks {
             let traces = self.trace_block_with(
                 num.into(),
                 TracingInspectorConfig::default_parity(),
                 move |tx_info, inspector, res, _, _| {
                     if let Some(idx) = tx_info.index {
+                        if idx > highest_idx {
+                            return Ok(None);
+                        }
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:306]: this is a smol improvement, however this still traces the entire block, 
https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/trace.rs#L353-L362

File: crates/rpc/rpc/src/trace.rs [hunk 3]
```diff
@@ -276,29 +276,34 @@ where
...
         }
 
-        // TODO: this could be optimized to only trace the block until the highest matching index in
-        // that block
-
         // trace all relevant blocks
         let mut block_traces = Vec::with_capacity(target_blocks.len());
-        for (num, indices) in target_blocks {
+        for (num, indices, highest_idx) in target_blocks {
             let traces = self.trace_block_with(
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:299]: there's something else we need to do here:
skipping blocks with empty indices entirely

**Review by mattsse — CHANGES_REQUESTED** [review]
that's a great start.

I have some nits,
the biggest improvement would still be to only trace the block until the highest index, for that we'd need a new function that is similar to:
https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/trace.rs#L363-L369
pub only traces the block until a given transaction, 
we can unify some code by moving this loop to a standalone function so it can be reused:

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/trace.rs#L408-L408

**Comment by smatthewenglish** [comment]
i wanted to localize this code in one place, instead of defining it twice, once for each function/
```
F: for<'a> Fn(
        TransactionInfo,
        TracingInspector,
        ExecutionResult,
        &'a revm_primitives::State,
        &'a CacheDB<StateProviderDatabase<StateProviderBox<'a>>>,
    ) -> EthResult<R>
    + Send
    + 'static,
```

so i defined it like this at the top of the file

```
trait TransactionCallback<R> 
where
    Self: for<'a> Fn(
        TransactionInfo,
        TracingInspector,
        ExecutionResult,
        &'a State,
        &'a CacheDB<StateProviderDatabase<StateProviderBox<'a>>>,
    ) -> EthResult<R>
    + Send
    + 'static,
{
}

impl<F, R> TransactionCallback<R> for F
where
    F: for<'a> Fn(
        TransactionInfo,
        TracingInspector,
        ExecutionResult,
        &'a State,
        &'a CacheDB<StateProviderDatabase<StateProviderBox<'a>>>,
    ) -> EthResult<R>
    + Send
    + 'static,
{
}
```

but i think that's still pretty redundant. I'm not that familiar with rust, it's why I wanna become a contributor to this project, to learn rust. is there a more idiomatic way to do that^?

File: crates/rpc/rpc/src/trace.rs [hunk 4]
```diff
@@ -401,6 +439,12 @@ where
                 let mut transactions = transactions.into_iter().enumerate().peekable();
 
                 while let Some((idx, tx)) = transactions.next() {
+                    // Check if current index exceeds the highest_index and break if it does
+                    if let Some(highest) = highest_index {
+                        if idx as u64 > highest {
+                            break;
+                        }
+                    }
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:447]: I'd like to move this to `transactions.into_iter().take()` instead

File: crates/rpc/rpc/src/trace.rs [hunk 5]
```diff
@@ -310,6 +338,7 @@ where
                         .into_localized_transaction_traces(tx_info);
                     Ok(Some(traces))
                 },
+                Some(highest_idx),
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:341]: this argument should be moved and should be the second argument, so that the closure is always the last argument

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, almost there

File: crates/rpc/rpc/src/trace.rs [hunk 6]
```diff
@@ -462,13 +457,9 @@ where
                     let ResultAndState { result, state } = res;
                     results.push(f(tx_info, inspector, result, &state, &db)?);
 
-                    // need to apply the state changes of this transaction before executing the
-                    // next transaction
-                    if transactions.peek().is_some() {
-                        // need to apply the state changes of this transaction before executing
-                        // the next transaction
-                        db.commit(state)
-                    }
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:427]: we still need this

File: crates/rpc/rpc/src/trace.rs [hunk 7]
```diff
@@ -39,6 +39,33 @@ pub struct TraceApi<Provider, Eth> {
...
+            TransactionInfo,
+            TracingInspector,
+            ExecutionResult,
+            &'a State,
+            &'a CacheDB<StateProviderDatabase<StateProviderBox<'a>>>,
+        ) -> EthResult<R>
+        + Send
+        + 'static
+{
+}
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:67]: I don't really like this because this is just a hack to make signatures nicer and not really needed

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]
nice, lgtm

**Pull request merged**